### ISSUE TITLE
No character validation in CDATA tags

### ISSUE THREAD

**Issue by ender672** [jimweirich/builder#2]
Builder can generate invalid XML when creating a CDATA tag with a string that contains characters that XML considers invalid (per http://www.w3.org/TR/REC-xml/#charsets).

This example will create invalid XML:

<pre>xml = Builder::XmlMarkup.new
xml.cdata!("\x00")</pre>

We should allow separate utf8 and XML Control Character (ampersand and brackets) escaping in String#to_xs.

<pre>diff --git a/lib/builder/xchar.rb b/lib/builder/xchar.rb
index 4578c2a..220d23f 100644
--- a/lib/builder/xchar.rb
+++ b/lib/builder/xchar.rb
@@ -92,10 +92,17 @@ class Fixnum
   # XML escaped version of chr. When <tt>escape</tt> is set to false
   # the CP1252 fix is still applied but utf-8 characters are not
   # converted to character entities.
-  def xchr(escape=true)
+  def xchr(escape_utf8=true, escape_markup_delimiters=true)
     n = XChar::CP1252[self] || self
     case n when *XChar::VALID
-      XChar::PREDEFINED[n] or (n&lt;128 ? n.chr : (escape ? "&##{n};" : [n].pack('U*')))
+      if escape_markup_delimiters && XChar::PREDEFINED.key?(n)
+        XChar::PREDEFINED[n]
+      else
+        if n&lt;128 then n.chr
+        elsif escape_utf8 then "&##{n};"
+        else [n].pack('U*')
+        end
+      end
     else
       '*'
     end
@@ -111,8 +118,8 @@ class String
   # XML escaped version of to_s. When <tt>escape</tt> is set to false
   # the CP1252 fix is still applied but utf-8 characters are not
   # converted to character entities.
-  def to_xs(escape=true)
-    unpack('U*').map {|n| n.xchr(escape)}.join # ASCII, UTF-8
+  def to_xs(escape_utf8=true, escape_markup_delimiters=true)
+    unpack('U*').map {|n| n.xchr(escape_utf8, escape_markup_delimiters)}.join # ASCII, UTF-8
   rescue
     unpack('C*').map {|n| n.xchr}.join # ISO-8859-1, WIN-1252
   end
diff --git a/lib/builder/xmlmarkup.rb b/lib/builder/xmlmarkup.rb
index 8483bed..d904c2f 100644
--- a/lib/builder/xmlmarkup.rb
+++ b/lib/builder/xmlmarkup.rb
@@ -280,7 +280,7 @@ module Builder
     def _special(open, close, data=nil, attrs=nil, order=[])
       _indent
       @target << open
-      @target << data if data
+      @target << data.to_xs(false, false) if data
       _insert_attributes(attrs, order) if attrs
       @target << close
       _newline
diff --git a/test/test_markupbuilder.rb b/test/test_markupbuilder.rb
index cae34ea..67dce33 100644
--- a/test/test_markupbuilder.rb
+++ b/test/test_markupbuilder.rb
@@ -351,6 +351,11 @@ class TestSpecialMarkup < Test::Unit::TestCase
     @xml.cdata!("TEST&CHECK")
     assert_equal "<![CDATA[TEST&CHECK]]>\n", @xml.target!
   end
+
+  def test_cdata_with_invalid_characters
+    @xml.cdata!("\x00")
+    assert_equal "<![CDATA[*]]>\n", @xml.target!
+  end
 end
 
 class TestIndentedXmlMarkup < Test::Unit::TestCase
</pre>

**Comment by ender672** [comment]
This makes builder slower at generating documents with lots of data in CDATA tags.

**Comment by ender672** [comment]
I ran into a case where I have to generate a large XML file (100's of MB) with lots of CDATA tags and it requires the removal of invalid characters from CDATA tags.

The patch above made the process painfully slow. I worked around it by letting libxml-ruby do the CDATA tag escaping:

```
class Builder::XmlMarkup
  def cdata!(text)
    self << LibXML::XML::Node.new_cdata(text)
  end
end
```

P.S. The original patch changes the internally used method xchr. The method call xchar(false, false) now means: "XMLize this string, but don't escape utf8 chars or xml delimiters."
